### DISCUSSION TITLE
Global state with State Crate

### DISCUSSION THREAD

**Discussion by wyhinton** [fltk-rs/fltk-rs#741] [Show and tell]
Hey all, 

Just wanted to share a app I've added to my[ FLTK examples repo.](https://github.com/wyhinton/FLTK-RS-Examples). Here we use a global Hashmap created with [state](https://github.com/sergiobenitez/state), to store, retrieve, and mutate values associated with a widget in a way similar to Javascript.

Here's a simple example using two counters:

```rust
use fltk::{app, button::Button, group::*, frame::Frame, enums::*, group::Pack, prelude::*, window::Window};
use state::Storage;
use std::collections::HashMap;
use std::sync::Mutex;

//convience macro for extracting value of an enum
#[macro_use]
macro_rules! as_variant {
    ($value:expr, $variant:path) => {
        match $value {
            $variant(x) => Some(x),
            _ => None,
        }
    };
}

static GLOBAL_MAP: Storage<Mutex<HashMap<String, WidgetValue>>> = Storage::new();

// enum wrapper for differnet value types allows our global map to store multiple types
#[derive(Clone, Copy)]
pub enum WidgetValue{
    Integer32(i32),
    Unsized32(u32)
    // String(String)
}

#[derive(Clone)]
pub struct Counter {
    count: Storage<i32>
}

//for a simple example we'll use global events and app::handle_main to trigger updates outside of the wrapped widget
pub struct GlobalEvents{}

impl GlobalEvents{
    pub const COUNTER_UPDATE: i32 = 41;
}

impl Counter {
    pub fn new(val: i32, id: &'static str) -> Self {
        //insert counters value into the global hashmap, using it's id as a key
        let mut map = GLOBAL_MAP.get().lock().unwrap();
        map.insert(id.into(),  WidgetValue::Integer32(0));


        let mut pack = Pack::new(0,0,100,150, None);
        pack.set_frame(FrameType::BorderFrame);
        // pack.set_color(Color::Black);
        //increment button
        let mut but_inc = Button::default().with_size(0, 40).with_label("+");
        but_inc.set_color(Color::Green);



        //frame for displaying current counter
                //fetch counter starting value
        let display_label = match map.get(id.into()).unwrap(){
            WidgetValue::Integer32(val)=>{
                let my_val = val.to_string();
                my_val
            }
            _=>("no value".to_string())
        };
        let mut frame = Frame::default()
            .with_size(0, 40)
            .with_label(&display_label.to_string());
        frame.set_frame(FrameType::FlatBox);
        frame.set_color(Color::Dark2);
        frame.set_label_color(Color::White);

        //decrement button
        let mut but_dec = Button::default().with_size(0, 40).with_label("-");
        but_dec.set_color(Color::Red);

        pack.end();

        //increment value, update frame label
        let mut frame_cl = frame.clone();
        but_inc.set_callback(move|_|{
            let enum_innner_val = *GLOBAL_MAP.get().lock().unwrap().get(id).unwrap();
            use WidgetValue::*;
            let inner_value = match enum_innner_val {
                Integer32(val)=>{
                    val
                },
                _=>(0)
            };
            frame_cl.set_label(&inner_value.to_string());
            *GLOBAL_MAP.get().lock().unwrap().get_mut(id).unwrap() = Integer32(inner_value+1);
            let _ = app::handle_main(GlobalEvents::COUNTER_UPDATE);
        });

        //decrement value, update frame label
        let mut frame_cl_2 = frame.clone();
        but_dec.set_callback(move|_|{
            let enum_innner_val = GLOBAL_MAP.get().lock().unwrap().get(id).unwrap().clone();
            use WidgetValue::*;
            let inner_value = match enum_innner_val {
                Integer32(val)=>{
                    val
                },
                _=>(0)
            };
            frame_cl_2.set_label(&inner_value.to_string());
            *GLOBAL_MAP.get().lock().unwrap().get_mut(id).unwrap() = Integer32(inner_value-1);
            let _ = app::handle_main(GlobalEvents::COUNTER_UPDATE);
        });
    
        Counter {
            count: Storage::from(val)
        }
    
    }
}

fn main(){
    //initialize store
    let initial_map = HashMap::new();
    GLOBAL_MAP.set(Mutex::new(initial_map));

    let app = app::App::default();
    app::swap_frame_type(FrameType::FlatBox);

    let mut wind = Window::default().with_size(500, 300).with_label("Counter");
    let mut container = Pack::new(150,100, 200, 200, None);
    container.set_spacing(20);
    container.set_type(PackType::Horizontal);

    let _c1 = Counter::new(0, "Counter 1");
    let _c2 = Counter::new(10, "Counter 2");
    container.end();    
    
    let label_pack= Pack::new(100,60,300,40,"App State");
    let mut state_frame= Frame::default().with_label("Click a Counter!").with_size(300,20).with_pos(100, 20);

    //update our "App State" frame whenever the counter buttons are presssed
    state_frame.handle(move |widg, ev| 
      if ev == GlobalEvents::COUNTER_UPDATE.into(){
          let counter_1_val = as_variant!(*GLOBAL_MAP.get().lock().unwrap().get("Counter 1").unwrap(), WidgetValue::Integer32).expect("failed to get int");
          let counter_2_val = as_variant!(*GLOBAL_MAP.get().lock().unwrap().get("Counter 2").unwrap(), WidgetValue::Integer32).expect("failed to get int");
          let state_str = format!("Counter 1 Val: {}, Counter 2 Val: {}", counter_1_val, counter_2_val);
          widg.set_damage(true);
          widg.set_label(&state_str);
        true
    } else {
        false
    });
    state_frame.set_frame(FrameType::FlatBox);
    state_frame.set_color(Color::Green);
    label_pack.end();
    wind.end();
    wind.show();
    app.run().unwrap();
}
```

Would welcome any feedback.

**Comment by eadf** [comment]
Nice! I got some notes:
* Introducing Mutex to an inherently single-threaded app seems overkill (and confusing), you could use `state::LocalStorage` and` Rc<RefCell>` instead.
* you could use `if let Some(MyEnum(value)) =` for extracting value of an enum. In an example it might be worth the effort to keep things easy to understand and read - i.e no macro magic.
* To make it visually clear that we are not cloning the entire object, but rather only cloning the `Rc` - I prefer to write  `Rc::clone(frame)` instead of `frame.clone()` (for example)

**Reply by MoAlyousef** [reply]
Thanks for exploring options for state management. 
It seems `state` has several advantages over lazy_static, I have yet to look closer at the code. A primary run shows the first click appears to be ignored, it might be from returning 0 to inner_value, or maybe the empty initialized GLOBAL_MAP.

**Reply by wyhinton** [reply]
>Introducing Mutex to an inherently single-threaded app seems overkill (and confusing), you could use state::LocalStorage and ?>Rc<RefCell> instead.

I agree this is a little confusing. Unfortunately with the state crate (also [event-emitter-rs](https://crates.io/crates/event-emitter-rs)), I'm finding a frequent requirement for ```Send+Sync```, which is incompatible with ```Rc<RefCell<>>```. See [```Storage``` from the state crate](https://docs.rs/state/0.5.2/state/struct.Storage.html):
```
impl<T: Send + Sync> Storage<T>
```
___

> you could use if let Some(MyEnum(value)) = for extracting value of an enum. In an example it might be worth the effort to keep things easy to understand and read - i.e no macro magic.

Yes I'm a little torn about rust macros. They can make debugging extremely difficult, but also remove boilerplate clutter. Is this what you were imagining?
```rust
  let mut counter_1_val = 0;
  if let Some(WidgetValue::Integer32(val)) = GLOBAL_MAP.get().lock().unwrap().get("Counter 1"){
  counter_1_val = *val;
  }
  let mut counter_2_val = 0;
  if let Some(WidgetValue::Integer32(val)) = GLOBAL_MAP.get().lock().unwrap().get("Counter 2"){
  counter_2_val = *val;
```
---

> To make it visually clear that we are not cloning the entire object, but rather only cloning the Rc - I prefer to write Rc::clone(frame) instead of frame.clone() (for example)

I agree that visually ```Rc::clone``` is more legible. In this instance are we creating a new Rc from the frame, then cloning that Rc?

```rust
let mut frame_cl_2 = Rc::clone(&Rc::new(frame));
```

**Reply by eadf** [reply]
> I agree this is a little confusing. Unfortunately with the state crate (also [event-emitter-rs](https://crates.io/crates/event-emitter-rs)), I'm finding a frequent requirement for `Send+Sync`, which is incompatible with `Rc<RefCell<>>`. See [`Storage` from the state crate](https://docs.rs/state/0.5.2/state/struct.Storage.html):

I was thinking about [`state::LocalStorage`](https://docs.rs/state/0.5.2/state/struct.LocalStorage.html), but I now see that it also requires `Send`- instantly disqualifying `Rc`  and `RefCell` types. It seems like the only way to get mutable LocalStorage data is through `std::cell:Cell`. - Not at all suitable for an entire hashmap. But maybe the individual elements inside the hashmap could be `std::cell:Cell`? 
Nope, the `Counter::new()` constructor  would not be able to insert new items.

I've not investigated event-emitter-rs, but I'm sure there are variants that does not require Sync.

>Yes I'm a little torn about rust macros. They can make debugging extremely difficult, but also remove boilerplate clutter. Is this what you were imagining?

Yeah, or something like this to emulate the  `.expect()`
```rust
let global_map = GLOBAL_MAP.get().lock().unwrap();
let counter_1_val = if let Some(WidgetValue::Integer32(v)) = global_map.get("Counter 1") {
    v
} else {
    panic!("failed to get int"); // equivalent of .expect("failed to get int");
};
let counter_2_val = if let Some(WidgetValue::Integer32(v)) = global_map.get("Counter 2") {
    v
} else {
    panic!("failed to get int"); // equivalent of .expect("failed to get int");
};
```

or if we define a `From` :
```rust
  let global_map = GLOBAL_MAP.get().lock().unwrap();
  let counter_1_val = i32::from(global_map.get("Counter 1").unwrap()); // From trait defines .from()..
  let counter_2_val:i32 = global_map.get("Counter 2").unwrap().into(); // and .into()
```

This makes the event logic easy to read, but requires a little more boiler-plate elsewhere, and it will have to be repeated for u32. :/
```rust
impl From<&WidgetValue> for i32 {
    fn from(op:&WidgetValue) -> i32 {
        if let WidgetValue::Integer32(v) = op {
            *v
        } else {
            panic!("failed to get int"); // equivalent of .expect("failed to get int");
        }
    }
}
```
>I agree that visually Rc::clone is more legible.

Hm, turns out that frame isn't a Rc in this context, so Rc::clone(frame) does not work here.  
Have not given frame.clone() a thought before, but what kind of wizardry is that? Two writable 'clones' referring to the same 'thing'  at two different places? :)

I guess we can put frame inside a Rc<RefCell<>> and pretend that it works like any other Rust object:
```rust
    let frame = Rc::from(RefCell::from(Frame::default()
                                           .with_size(0, 40)
                                           .with_label(&display_label.to_string())));
    frame.borrow_mut().set_frame(FrameType::FlatBox);
    frame.borrow_mut().set_color(Color::Dark2);
    frame.borrow_mut().set_label_color(Color::White);
    ...
    but_inc.set_callback({
        let frame = Rc::clone(&frame);
        move|_|{
            let enum_innner_val = *GLOBAL_MAP.get().lock().unwrap().get(id).unwrap();
            ....
            frame.borrow_mut().set_label(&inner_value.to_string());
```
